make.py

```diff
@@ -0,0 +1,341 @@
+#!/usr/bin/env python3
+import os
+import subprocess
+import shutil
+import sys
+from pathlib import Path
+
+HELP = f"""Available targets:
+- help              displays this message and exits
+- install           installs the project
+- uninstall         uninstalls the project
+- clean             removes build artifacts from the project directories
+- build-sql         constructs the sql files for the extension
+- clean-sql         removes sql file artifacts from the project directories
+- install-sql       installs the sql files into the postgres installation
+- uninstall-sql     removes the sql extension from the postgres installation
+- install-prior-py  installs the python package and dependencies from previously released versions
+- install-py        installs the python package and dependencies for the current version
+- clean-py          removes python build artifacts from the project directories
+- uninstall-py      removes the python package and dependencies from the system
+- test              runs the unit tests in the docker database
+- docker-build      builds the development docker image
+- docker-run        launches a container in docker using the docker image
+- docker-stop       stops the container
+- docker-rm         deletes the development container"""
+
+
+def versions() -> list[str]:
+    return ["0.4.0", "0.3.0", "0.2.0", "0.1.0"]  # ADD NEW VERSIONS TO THE FRONT OF THIS LIST! STAY SORTED PLEASE
+
+
+def this_version() -> str:
+    return versions()[0]
+
+
+def prior_versions() -> list[str]:
+    return versions()[1:] if len(versions()) > 1 else []
+
+
+def project_dir() -> Path:
+    return Path(__file__).resolve().absolute().parent
+
+
+def sql_dir() -> Path:
+    return project_dir().joinpath("sql")
+
+
+def src_dir() -> Path:
+    return project_dir().joinpath("src")
+
+
+def incremental_sql_dir() -> Path:
+    return sql_dir().joinpath("incremental")
+
+
+def idempotent_sql_dir() -> Path:
+    return sql_dir().joinpath("idempotent")
+
+
+def idempotent_sql_files() -> list[Path]:
+    paths = [x for x in idempotent_sql_dir().glob("*.sql")]
+    paths.sort()
+    return paths
+
+
+def incremental_sql_files() -> list[Path]:
+    paths = [x for x in incremental_sql_dir().glob("*.sql")]
+    paths.sort()
+    return paths
+
+
+def output_sql_file() -> Path:
+    return project_dir().joinpath(f"sql/ai--{this_version()}.sql")
+
+
+def build_incremental_sql_file(input_file: Path) -> str:
+    template = sql_dir().joinpath("migration.sql").read_text()
+    migration_name = input_file.name
+    migration_body = input_file.read_text()
+    version = this_version()
+    return template.format(migration_name=migration_name, migration_body=migration_body, version=version)
+
+
+def build_idempotent_sql_file(input_file: Path) -> str:
```

Comment:
If `this_version` is always the current version, couldn't we extract this code into its own extension function? All the other functions would just call this new other function instead of having to do a search and replace?

Reply:
We *could make it another SQL function, but I was worried about the overhead of making another extra SQL call for EVERY function call. Maybe that's an unnecessary concern. I didn't test the impact of the overhead.